src/executable_code_finder/lib.rs

```diff
@@ -0,0 +1,241 @@
+                        }
+                        Ok(res) => {
+                            let mut executable_lines = executable_lines.lock().unwrap();
+                            if !res.executable_lines.is_empty() {
+                                executable_lines.push(res);
+                            }
+                        }
+                    };
+
+                    let mut tally = files_processed.lock().unwrap();
+                    *tally += 1;
+
+                    update_progressbar(*tally, bar.clone());
+                }
+            });
+
+            handles.push(handle);
+        }
+
+        for handle in handles {
+            handle.join().unwrap();
+        }
+
+        if let Some(bar) = &bar {
+            bar.finish_and_clear();
+        }
+
+        if logger.can_log_timing() {
+            logger.log_sync(&format!(
+                "Scanning files took {:.2?}",
+                file_scanning_now.elapsed()
+            ));
+        }
+    }
+
+    let _invalid_files = Arc::try_unwrap(invalid_files)
+        .unwrap()
+        .into_inner()
+        .unwrap();
+
+    Ok(Arc::try_unwrap(executable_lines).unwrap().into_inner().unwrap())
+}
+
+fn get_filesystem(
+    files_to_scan: &mut Vec<String>,
+    interner: &mut Interner,
+    logger: &Logger,
+    scan_dirs: &Vec<String>,
+    existing_file_system: &Option<VirtualFileSystem>,
+    config: &Arc<Config>,
+    cache_dir: Option<&String>,
+    files_to_analyze: &mut Vec<String>,
+) -> VirtualFileSystem {
+    let mut file_system = VirtualFileSystem::default();
+
+    add_builtins_to_scan(files_to_scan, interner, &mut file_system);
+
+    logger.log_sync("Looking for Hack files");
+
+    for scan_dir in scan_dirs {
+        logger.log_debug_sync(&format!(" - in {}", scan_dir));
+
+        files_to_scan.extend(file_system.find_files_in_dir(
+            scan_dir,
+            interner,
+            existing_file_system,
+            config,
+            cache_dir.is_some() || config.ast_diff,
+            files_to_analyze,
+        ));
+    }
+
+    file_system
+}
+
+
+fn update_progressbar(percentage: u64, bar: Option<Arc<ProgressBar>>) {
+    if let Some(bar) = bar {
+        bar.set_position(percentage);
+    }
+}
+
+pub(crate) fn scan_file(
+    interner: &ThreadedInterner,
+    root_dir: &str,
+    file_path: FilePath,
+    logger: &Logger,
+) -> Result<ExecutableLines, ParserError>{
+    let interner = interner
+        .parent
+        .lock()
+        .unwrap();
+    let str_path = interner
+        .lookup(&file_path.0)
+        .to_string();
+
+    logger.log_debug_sync(&format!("scanning {}", str_path));
+
+    let aast = hakana_workhorse::get_aast_for_path(file_path, &str_path);
+
+    let aast = match aast {
+        Ok(aast) => aast,
+        Err(err) => {
+            return Err(err);
+        }
+    };
+
+    let mut checker = Scanner {
+    };
+
+    let mut context= Vec::new();
+    match visit(&mut checker, &mut context, &aast.0) {
```

Comment:
you should be able to map the output instead of using `match`.

so `visit(...).map(|_| ExecutableLines { ... })`

Reply:
I went back and forth between these, but now I moved the panic! into this method, so came back to match